### PR TITLE
[DOCU-2413] and [DOCU-2447] Performance Testing Framework IA rework

### PR DESCRIPTION
**Pull request by cloudjumpercat**
### Summary

- New How-to Doc: Contains the following sections from this [doc](https://docs.konghq.com/gateway/latest/plan-and-deploy/performance-testing-framework/): Installation Drivers Workflow Sample Output section
- Contains the reference portions of this [doc](https://docs.konghq.com/gateway/latest/plan-and-deploy/performance-testing-framework/)

### Reason

https://konghq.atlassian.net/browse/DOCU-2413
https://konghq.atlassian.net/browse/DOCU-2447

### CODE REVIEW

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 1]
```diff
@@ -1,7 +1,236 @@
...
+The framework uses [busted](https://olivinelabs.com/busted/) and some
+Lua development dependencies of Kong. To set up the environment,
+run `make dev` on your Kong repository to install all Lua dependencies.
+
+## Usage
+
+### Basic usage
+
+Install Lua development dependencies of Kong being installed and invoke each test file with `busted`. The following runs the simple RPS and latency
+test using `docker` driver.
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:24]:
```suggestion
test using the`docker` driver.
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 2]
```diff
@@ -1,7 +1,236 @@
...
+
+The framework uses [busted](https://olivinelabs.com/busted/) and some
+Lua development dependencies of Kong. To set up the environment,
+run `make dev` on your Kong repository to install all Lua dependencies.
+
+## Usage
+
+### Basic usage
+
+Install Lua development dependencies of Kong being installed and invoke each test file with `busted`. The following runs the simple RPS and latency
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:23]:
```suggestion
Install Lua development dependencies of Kong being installed and invoke each test file with `busted`. The following runs the RPS and latency
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 3]
```diff
@@ -1,7 +1,236 @@
...
+test using `docker` driver.
+
+```shell
+PERF_TEST_USE_DAILY_IMAGE=true PERF_TEST_VERSIONS=git:master,git:perf/your-other-branch bin/busted -o gtest spec/04-perf/01-rps/01-simple_spec.lua
+```
+
+User can also run the performance test framework inside a Kong container from scratch, from other development tools
+like [gojira](https://github.com/kong/gojira) or [Pongo](https://github.com/kong/kong-pongo), if they don't have development environment.
+Note to use the `docker` driver inside a container, the Docker daemon socket `/var/run/docker.sock` needs to be mounted inside the container.
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:32]:
```suggestion
```

Comment (Guaris) [reply src/gateway/understanding-kong/kong-performance-testing.md:32]: I learned last week that we don't want to steer customers towards Gojira as a solution.

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 4]
```diff
@@ -1,7 +1,236 @@
...
+
+### Terraform managed instances
+
+By default, terraform doesn't teardown instances after each test in `lazy_teardown`; this allows user
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:36]:
```suggestion
By default, terraform doesn't teardown instances after each test in `lazy_teardown`; allowing users
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 5]
```diff
@@ -1,7 +1,236 @@
...
+to run multiple tests consecutively without rebuilding the infrastructure each time.
+
+This behaviour can be changed by setting `PERF_TEST_TEARDOWN_ALL` environment variable to true. User can
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:39]:
```suggestion
This behaviour can be changed by setting the `PERF_TEST_TEARDOWN_ALL` environment variable to true. Users can
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 6]
```diff
@@ -1,7 +1,236 @@
...
+also manually teardown the infrastructure by running:
+
+```
+PERF_TEST_TEARDOWN_ALL=1 PERF_TEST_DRIVER=terraform PERF_TEST_TERRAFORM_PROVIDER=your-provider bin/busted -o gtest -t single spec/04-perf/99-teardown/
+```
+
+### AWS
+
+When using `terraform` driver and `aws-ec2` as provider, AWS credentials can be provided by environment variables.
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:48]:
```suggestion
When using the `terraform` driver and `aws-ec2` as the provider, AWS credentials can be provided using environment variables.
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 7]
```diff
@@ -1,7 +1,236 @@
...
+It's common to pass `AWS_PROFILE` to point a stored AWS credentials profile, or `AWS_ACCESS_KEY_ID` and
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:49]:
```suggestion
It's common to pass the `AWS_PROFILE` variable to point to a stored AWS credentials profile, or `AWS_ACCESS_KEY_ID` and
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 8]
```diff
@@ -1,7 +1,236 @@
...
+`AWS_SECRET_ACCESS_KEY` to use credential directly. Please refer to the
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:50]:
```suggestion
`AWS_SECRET_ACCESS_KEY` to use credentials directly. Please refer to the
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 9]
```diff
@@ -1,7 +1,236 @@
...
+
+
+## Drivers
+
+Three "drivers" are implemented depending on different environments, accuracy
+requirements, and setup complexity.
+
+| Driver   | Test between git commits | Test between binary releases | Flame Graph | Test unreleased version    |
+|----------|--------------------------|------------------------------|------------|----------------------------|
+| docker   | yes                      | yes                          |            | yes (`use_daily_image` = true) |
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:76]:
```suggestion
| Docker   | yes                      | yes                          |            | yes (`use_daily_image` = true) |
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 10]
```diff
@@ -1,7 +1,236 @@
...
+|terraform | yes                      | yes                          | yes        | yes (`use_daily_image` = true) |
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:77]:
```suggestion
|Terraform | yes                      | yes                          | yes        | yes (`use_daily_image` = true) |
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 11]
```diff
@@ -1,7 +1,236 @@
...
+
+You must install the following Lua development dependencies to use either of the drivers:
+
+- **docker** Driver is solely based on Docker images. It's the most convenient driver to setup
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:81]:
```suggestion
- **Docker** Driver is based on Docker images.
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 12]
```diff
@@ -1,7 +1,236 @@
...
+as it requires less dependencies. But it may also be influenced by other local programs
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:82]:
```suggestion
It requires fewer dependencies.
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 13]
```diff
@@ -1,7 +1,236 @@
...
+and Docker network performance. And it doesn't support FlameGraph generation.
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:83]:
```suggestion
And it doesn't support FlameGraph generation.
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 14]
```diff
@@ -1,7 +1,236 @@
...
+
+- **terraform** Driver runs in remote VM or bare metal machine. It's most accurate,
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:85]:
```suggestion
- **terraform** Driver runs in a remote VM or on a bare metal machine. It's most accurate,
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 15]
```diff
@@ -1,7 +1,236 @@
...
+but it requires Terraform knowledge to operate and setup.
+
+    * Requires the [Terraform](https://www.terraform.io/downloads.html) binary be installed.
+    * Requires git binary if testing between git commits. When testing between git commits,
+    the framework assumes the current directory is Kong's repo. It will stash your working
+    directory and remove the stash after the test is finished. When using the docker or Terraform driver,
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:91]:
```suggestion
    directory and remove the stash after the test is finished. When using the Docker or Terraform driver,
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 16]
```diff
@@ -1,7 +1,236 @@
...
+    the framework derives the base version of each commit and uses the matching Docker image or
+    Kong binary package and puts local source code inside.
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:93]:
```suggestion
    Kong binary package and puts the local source code inside.
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 17]
```diff
@@ -1,7 +1,236 @@
...
+
+## Workflow
+
+Like Kong's integration tests, the performance test is written in Lua. Several
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:97]:
```suggestion
Like Kong's integration tests, the performance test is written in Lua. Several
```
```suggestion
Multiple
```

File: src/gateway/understanding-kong/kong-performance-testing.md [hunk 18]
```diff
@@ -1,7 +1,236 @@
...
+
+      print(("### Combined result for Kong %s:\n%s"):format(version, assert(perf.combine_results(results))))
+
+      perf.save_error_log("output/" .. version:gsub("[:/]", "#") .. "-single_route.log")
+    end)
+  end)
+end
+```
+
+The test can be run as a normal busted-based test. Run it with `bin/busted path/to/this_file_spec.lua`.
```
Comment (Guaris) [inline src/gateway/understanding-kong/kong-performance-testing.md:177]:
```suggestion
The test can be run as a normal Busted test, using `bin/busted path/to/this_file_spec.lua`.
```

**Review by Guaris — CHANGES_REQUESTED** [review]

**Comment by cloudjumpercat** [comment]
Not sure why Vale is failing. I don't get any errors when I run it locally.

**Comment by Guaris** [comment]
> Not sure why Vale is failing. I don't get any errors when I run it locally.

It's from the terraform variables. You can either wrap them in code tags or ignore the section. Both are okay solutions.

**Review by Guaris — APPROVED** [review]

**Pull request merged**